test_conformance/subgroups/test_workgroup_ballot.cpp

```diff
@@ -0,0 +1,973 @@
+                                                       : bit_value = 0;
+                    if (i >= 64 && i < 96)
+                        (mx[ii + i].s2 & bit_mask) > 0 ? bit_value = 1
+                                                       : bit_value = 0;
+                    if (i >= 96 && i < 128)
+                        (mx[ii + i].s3 & bit_mask) > 0 ? bit_value = 1
+                                                       : bit_value = 0;
+
+                    bit_value == 1 ? tr = { 1, 0, 0, 0 } : tr = { 0, 0, 0, 0 };
+
+                    rr = my[ii + i];
+                    if (!compare(rr, tr))
+                    {
+                        log_error(
+                            "ERROR: sub_group_%s mismatch for local id %d in "
+                            "sub group %d in group %d obtained {%d, %d, %d, "
+                            "%d}, expected {%d, %d, %d, %d}\n",
+                            operation_names(operation), i, j, k, rr.s0, rr.s1,
+                            rr.s2, rr.s3, tr.s0, tr.s1, tr.s2, tr.s3);
+                        return -1;
+                    }
+                }
+            }
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+
+        return 0;
+    }
+};
+
+
+// Test for bit count/inclusive and exclusive scan/ find lsb msb ballot function
+template <typename Ty, BallotOp operation> struct BALLOT3
+{
+    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+    {
+        int i, ii, j, k, n;
+        int nj = (nw + ns - 1) / ns;
+        int e;
+        ii = 0;
+        for (k = 0; k < ng; ++k)
+        { // for each work_group
+            for (j = 0; j < nj; ++j)
+            { // for each subgroup
+                ii = j * ns;
+                n = ii + ns > nw ? nw - ii : ns;
+                if (operation == BallotOp::ballot_bit_count
+                    || operation == BallotOp::ballot_inclusive_scan
+                    || operation == BallotOp::ballot_exclusive_scan)
+                {
+                    // Initialize data matrix indexed by local id and sub group
+                    // id
+                    e = (int)(genrand_int32(gMTdata) % 3);
+                    switch (e)
+                    {
+                        case 0: memset(&t[ii], 0, n * sizeof(Ty)); break;
+                        case 1:
+                            memset(&t[ii], 0, n * sizeof(Ty));
+                            i = (int)(genrand_int32(gMTdata) % (cl_uint)n);
+                            set_value(t[ii + i], 41);
+                            break;
+                        case 2: memset(&t[ii], 0xff, n * sizeof(Ty)); break;
+                    }
+                }
+                else if (operation == BallotOp::ballot_find_lsb
+                         || operation == BallotOp::ballot_find_msb)
+                {
+                    // Regarding to the spec, find lsb and find msb result is
```

Comment:
This comment is slightly incorrect I believe. The value needs at least one bit set for a work-item in the subgroup. If the subgroup is very small and no bit is set for these work-items, this would also be UB.

Maybe another way to generate valid data would be to:
 1. pick a random number `x` (the domain is not important),
 2. pick a random value `i` in the domain `[0, subgroup_size]`.
 3. set the `i`-th bit of `x` to `1`.
 4. set `t[ii]` to `x`.

What is your opinion?

Reply:
Spec says that in case of no bit set (zeros on all bits in value) then behaviour is undefined. I've simplified a little bit generation values. Now it calculates values for each work item in the subgroup:
```
char x = (genrand_int32(gMTdata)) & 0xff;
// undefined behaviour in case of 0;
x = x ? x : 1;
memset(&t[wg_offset + wi_id], x, sizeof(Ty));
```